donut/modules/marketplace/routes.py

```diff
@@ -0,0 +1,342 @@
+
+
+@blueprint.route('/marketplace/search')
+def query():
+    """Displays all results for the query in category category_id, which can be
+       'all' if no category is selected."""
+
+    category_id = flask.request.args["cat"]
+    query = flask.request.args["q"]
+
+    fields = ["cat_id", "item_title", "textbook_title", "item_price", "user_id", "item_timestamp"]
+    # Create a dict of the passed in attributes which are filterable
+    filterable_attrs = ["item_id", "cat_id", "user_id", "item_title",
+            "item_details", "item_images", "item_condition",
+            "item_price", "item_timestamp", "item_active",
+            "textbook_id", "textbook_isbn", "textbook_edition", "textbook_title"]
+    attrs = { tup:flask.request.args[tup]
+            for tup in flask.request.args if tup in filterable_attrs }
+    if category_id == "all":
+        category_id = 0
+    else:
+        attrs["cat_id"] = category_id
+        # only pass in the cat_id to get_marketplace_items_list_data if it's not
+        # "all", because cat_id (and everything in attrs) goes into a WHERE
+        # clause, and not specifying is the same as selecting all.
+
+    # now, the category id had better be a number
+    try:
+        cat_id_num = int(category_id)
+        (datalist, headers, links) = helpers.generate_search_table(fields=fields, attrs=attrs, query=query)
+
+        return helpers.render_with_top_marketplace_bar('search.html', datalist=datalist, cat_id=cat_id_num, headers=headers, links=links)
+
+    except ValueError:
+        # not a number? something's wrong
+        return flask.render_template('404.html')
+
+
+@blueprint.route('/marketplace/view_item')
+def view_item():
+    """View additional details about item <item_id>, passed through flask.request.args."""
+
+    if "item_id" not in flask.request.args:
+        return flask.render_template('404.html')
+
+    # make sure item_id is a number
+    try:
+        item_id = int(flask.request.args["item_id"])
+
+        stored = {}
+        storedFields = ["textbook_id", "item_id", "cat_id", "user_id", "item_title", "item_details", "item_condition", "item_price", "item_timestamp", "item_active", "textbook_edition", "textbook_isbn", "textbook_title", "textbook_author"]
+        data = helpers.get_table_list_data(['marketplace_items', 'marketplace_textbooks'], storedFields, {'item_id': item_id}, "NATURAL LEFT JOIN")[0]
+        for i in range(len(data)):
+            stored[storedFields[i]] = data[i]
+
+        # cat_title from cat_id
+        cat_id_map = {sublist[0]: sublist[1] for sublist in helpers.get_table_list_data("marketplace_categories", ["cat_id", "cat_title"])}
+        cat_title=cat_id_map[stored["cat_id"]]
+
+        # full_name and email from user_id
+        data = helpers.get_table_list_data(['members', 'members_full_name'], ['full_name', 'email'], {'user_id': stored['user_id']}, "NATURAL LEFT JOIN")[0]
```

Comment:
Since you use the `user_id` from stored, which comes from the first result,
this always uses the same `user_id` (not the `user_id` of the user who is logged in).
You should check if there's a helper function to do this in the core module and add it if there's not.